Add Stream.Shutdown() method

sheerun opened the pull request:
This is something like Wait() we discussed earlier, it waits for all readers to gracefully close, but prevents any new readers being created and doesn't remove the file afterwards

I've also updated the tests so they test every filesystem and additional one (slowFs) that has delays.

How to tests this:

1. checkout first commit and verify that that all past tests pass
2. checkout second commit and verify new tests pass that cover Shutdown method and more

sheerun commented:
One extra commit: Shutdown first Closes stream, and only then waits for readers to close, not the other way around.

sheerun commented:
My use case: Create writer and make it readable for 10 seconds, then Shutdown() it preventing new readers registering, but wait for another 60 seconds for existing readers to finish their job. After that Cancel() if Shutdown() haven't completed yet.

sheerun commented:
Should I create ErrShutdown instead of returing ErrCanceled for new readers after shutdown?

djherbis commented:
Sorry for the conflicts, I've actually added a simpler version of Shutdown imho, but didn't expose it yet. Let me know what you think.

I'm curious, what the slow fs is supposed to test?

sheerun commented:
It's OK, but please copy my tests on top of your implementation

SlowFS is just in case, I've just created it if there was some weird behavior connected with FS being slow, but it doesn't seems so. It doesn't hurt to leave it as well.

djherbis commented:
Can you try fast forwarding this branch from master? I merged it in, but GitHub didn't notice this is merged, I think because I did a rebase, I tried pushing my changes to the branch but I get permission denied.

sheerun commented:
@djherbis I see you've taken rather different route implementing this feature. I've rebased this branch to use your implementation but left one test that fails. ShutdownWithErr should allow any existing readers to finish reading, while your implementation blocks indefinitely on WaitForZeroHandles

In my implementation I've introduced another shutdownState, but also only waited for reader handles to finish, and then manually closed writer instead. It prevents infinite blocking.

sheerun commented:
If you'd like to see my implementation again I can upload it somewhere

djherbis commented:
My impl. still allows existing readers to finish, to gain the behaviour
yours had, just call Stream.Close() then shutdown.

Is the behaviour any diff if you do that?

On Thu, Jan 11, 2018, 12:38 AM Adam Stankiewicz <notifications@github.com>
wrote:

> If you'd like to see my implementation again I can upload it somewhere
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/djherbis/stream/pull/7#issuecomment-356863678>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABds6iHZxTLrbDzKFd-qtnXrHUCLdaAeks5tJchtgaJpZM4RWyd7>
> .
>

sheerun commented:
yeah, the test fails as well, even with close before shutdown.

btw. I don't see the point of shutdown without close so shutdown probably could do it

djherbis commented on stream_test.go:
```diff
@@ -371,34 +371,35 @@ func testShutdownAfterClose(t *testing.T, fs FileSystem) {
 
+	go func() {
+		time.Sleep(50 * time.Millisecond)
+		_, err := ioutil.ReadAll(r)
```
You never close this Reader, so ShutdownWithErr blocks forever.

sheerun replied:
Allright, in my implementation I close readers automatically when they read until the end of writer. I guess this design decision is up to you.

sheerun commented:
I've pushed updated test with closed reader, not much is changed, really. Thanks :)

djherbis commented:
The reason for not calling Close in Shutdown is to make it optional. I can see a case where maybe you want to signal that your shutting down before you've finished writing to the stream.

As for why it doesn't auto-close when you hit the end, that's because you still have ReadAt, maybe once you've read the whole stream you'll want to backup and re-reading a middle section.

djherbis commented on stream_test.go:
```diff
@@ -234,7 +234,7 @@ func testCloseUnblocksBlockingRead(t *testing.T, fs FileSystem) {
 		_, err := ioutil.ReadAll(r)
 		if err == nil || err == io.EOF {
-			t.Error("expected an error on a blocking Read for a closed Reader")
+			t.Error("expeceted an error on a blocking Read for a closed Reader")
```
expected*
